src/main/java/uk/ac/ebi/embl/converter/gff3/reader/GFF3FileReader.java

```diff
@@ -0,0 +1,176 @@
+                String name = m.group("name");
+                long start = Long.parseLong(m.group("start"));
+                long end = Long.parseLong(m.group("end"));
+                String score = m.group("score");
+                String strand = m.group("strand");
+                String phase = m.group("phase");
+                String attributes = m.group("attributes");
+                Map<String, String> attributesMap = attributesFromString(attributes);
+                Optional<String> id =
+                        attributesMap.containsKey("ID") ? Optional.of(attributesMap.get("ID")) : Optional.empty();
+                Optional<String> parentId = attributesMap.containsKey("Parent")
+                        ? Optional.of(attributesMap.get("Parent"))
+                        : Optional.empty();
+
+                GFF3Feature feature = new GFF3Feature(
+                        id, parentId, accession, source, name, start, end, score, strand, phase, attributesMap);
+                id.ifPresent(s -> featureIdx.put(s, feature));
+
+                if (parentId.isPresent()) {
+                    if (featureIdx.containsKey(parentId.get())) {
+                        featureIdx.get(parentId.get()).addChild(feature);
+                    }
+                    // TODO: Validation check, no parent found
+                } else if (attributesMap.containsKey("gene")) {
+                    String gene = attributesMap.get("gene");
+                    geneMap.putIfAbsent(gene, new ArrayList<>());
+                    List<GFF3Feature> geneFeatures = geneMap.get(gene);
+                    geneFeatures.add(feature);
+                } else {
+                    nonGeneFeatures.add(feature);
+                }
+                // TODO: Validation check accession matches sequence region.
+
+                continue;
+            }
+            break;
+        }
+
+        if (directives.isEmpty() && geneMap.isEmpty() && nonGeneFeatures.isEmpty()) {
+            return null;
+        }
+        return new GFF3Annotation(new GFF3Directives(directives), geneMap, nonGeneFeatures);
+    }
+
+    private static Map<String, String> attributesFromString(String line) {
+        Map<String, String> attributes = new HashMap<>();
+        String[] parts = line.split(";");
+        for (String part : parts) {
+            String[] keyValue = part.split("=");
+            attributes.put(keyValue[0], keyValue[1]);
+        }
+        return attributes;
+    }
+
+    private List<GFF3Annotation> readAnnotationList() throws IOException, GFF3ValidationError, GFF3ValidationError {
+        List<GFF3Annotation> annotations = new ArrayList<>();
+        while (true) {
+            // readAnnotation
+            GFF3Annotation annotation = readAnnotation();
+            if (annotation == null) {
+                break;
+            } else {
+                annotations.add(annotation);
+            }
+        }
+        return annotations;
```

Comment:
Suggestion to add the list of annotations in readAnnotation() itself and remove this method readAnnotationList().